Added support for async property access

sompylasar opened the pull request:
- Async JSON and JSONParse.
- Transform the provided source code to handle promises (via parsing
  and AST traversal).
- Blacklist JavaScript-specific code constructs (a bonus to AST
  traversal).
- Blacklist JavaScript-specific global variables (not via AST yet,
  TODO).
- Added inspect documentation to the API items (e.g. evaluating just `HTTP` will print its documentation).
- Treat `undefined` result as an error. TODO(sompylasar): Think up something for the APIs that return void.

sompylasar committed:
bricks-net-api: Added async JSON and JSONParse

sompylasar committed:
Print `undefined` if in terminal

sompylasar committed:
Treat `undefined` result as an error.

TODO(sompylasar): Think up something for the APIs that return void.

sompylasar committed:
bricks-timing: Handle multiple `end` calls.

sompylasar committed:
bricks-prettyprint: Added inspect documentation

Return from `inspect` on an object to provide documentation for the
object.

sompylasar committed:
bricks-net-api: Added inspect documentation to the API items

sompylasar committed:
bricks-prettyprint: Print documentation only to the terminal

Made it gray.

sompylasar committed:
Added support for async property access

* Transform the provided source code to handle promises (via parsing
and AST traversal).
* Blacklist JavaScript-specific code constructs (a bonus to AST
traversal).
* Blacklist JavaScript-specific global variables (not via AST yet,
TODO).

dkorolev commented on lib/bricks-evaluate.js:
```diff
@@ -5,25 +5,49 @@ var vm = require('vm');
-	delete context.require;
-	delete context.module;
-	delete context.global;
+	// - the standard obejct types that are not in `global`.
```
Typo, `s/obejct/object/`.

sompylasar replied:
Fixed.

dkorolev commented on lib/bricks-evaluate.js:
Do we have any `void`-returning methods in Bricks' API?

dkorolev commented on lib/bricks-timing.js:
```diff
@@ -5,6 +5,7 @@ module.exports = function () {
 		startTime: Date.now(),
 		endTime: undefined,
 		end: function () {
+			if (timing.endTime) { return timing; }
```
Looks scary. Isn't it generally an erroneous case?

sompylasar replied:
Will replace with a throw.

dkorolev commented:
Looks good, just a few minor comments.

Unfortunately, it also does not work for me:

```
dima@dima-i9:~/github/sompylasar/KnowSheet-REPL$ node run
KnowSheet> JSONParse(HTTP(GET("http://httpbin.org/get?test=1")).body)
--------------------------------------------------------------------------------
ReferenceError: GET is not defined
--------------------------------------------------------------------------------
(109ms)
```

```
$ echo 'JSONParse(HTTP(GET("http://httpbin.org/get?test=1")).body).args' | node run | cat
ReferenceError: GET is not defined
```

sompylasar committed:
s/obejct/object/

sompylasar commented on lib/bricks-evaluate.js:
These:
https://github.com/KnowSheet/Bricks/blob/master/net/api/impl/posix_server.h#L159
https://github.com/KnowSheet/Bricks/blob/master/net/api/impl/posix_server.h#L173
https://github.com/KnowSheet/Bricks/blob/master/net/api/impl/posix_server.h#L190

sompylasar committed:
bricks-timing: Throw if called twice

dkorolev commented on lib/bricks-evaluate.js:
We won't need `Join()` in REPL.

As for `Register()/UnRegister()`, I was thinking of converting them to a scoped version (return a handle from `Register()`, automatically un-register once that handle goes out of scope). What do you think, in general and REPL-wise?

sompylasar committed:
Added development mode for debugging

Usage: replace `node run` with `NODE_ENV=development node run` in the
command line.

sompylasar commented:
@dkorolev Very strange. Please post a log from running in development mode I've just added: replace `node run` with `NODE_ENV=development node run` in the command line.

dkorolev commented:
Here is is:

```
$ NODE_ENV=development node run
KnowSheet> JSONParse(HTTP(GET("http://httpbin.org/get?test=1")).body)
1 | (JSONParse(__when1019(HTTP(GET("http://httpbin.org/get?test=1"))).then(function(ret) {
2 |   return ret.body;
3 | }))
4 | )

--------------------------------------------------------------------------------
ReferenceError: GET is not defined
    at [object Object]:1:28
    at Script.(anonymous function) [as runInContext] (vm.js:41:22)
    at Function.evaluate (/home/dima/github/sompylasar/KnowSheet-REPL/lib/bricks-evaluate.js:197:23)
    at KnowSheetREPLServer.options.eval (/home/dima/github/sompylasar/KnowSheet-REPL/lib/repl.js:62:12)
    at Interface.<anonymous> (repl.js:239:12)
    at Interface.emit (events.js:95:17)
    at Interface._onLine (readline.js:202:10)
    at Interface._line (readline.js:531:8)
    at Interface._ttyWrite (readline.js:760:14)
    at ReadStream.onkeypress (readline.js:99:10)
------------------------------
```

sompylasar committed:
Added context display in development mode

sompylasar commented:
@dkorolev Once again please.

dkorolev commented:
```
$ NODE_ENV=development node run
KnowSheet> JSONParse(HTTP(GET("http://httpbin.org/get?test=1")).body)
{ ArrayBuffer: undefined,
  Int8Array: undefined,
  Uint8Array: undefined,
  Uint8ClampedArray: undefined,
  Int16Array: undefined,
  Uint16Array: undefined,
  Int32Array: undefined,
  Uint32Array: undefined,
  Float32Array: undefined,
  Float64Array: undefined,
  DataView: undefined,
  global: undefined,
  process: undefined,
  GLOBAL: undefined,
  root: undefined,
  Buffer: undefined,
  setTimeout: undefined,
  setInterval: undefined,
  clearTimeout: undefined,
  clearInterval: undefined,
  setImmediate: undefined,
  clearImmediate: undefined,
  console: undefined,
  Array: undefined,
  Boolean: undefined,
  Date: undefined,
  Function: undefined,
  Number: undefined,
  Object: undefined,
  RegExp: undefined,
  String: undefined,
  Error: undefined,
  EvalError: undefined,
  RangeError: undefined,
  ReferenceError: undefined,
  SyntaxError: undefined,
  TypeError: undefined,
  URIError: undefined,
  require: undefined,
  module: undefined,
  JSON: // KnowSheet JSON,
  JSONParse: // KnowSheet JSONParse,
  SaveResponseToFile: // KnowSheet SaveResponseToFile,
  KeepResponseInMemory: // KnowSheet KeepResponseInMemory,
  HTTP: // KnowSheet HTTP,
  POSTFromFile: // KnowSheet POSTFromFile,
  POST: // KnowSheet POST,
  GET: // KnowSheet GET,
  HTTPResponseCodeAsString: // KnowSheet HTTPResponseCodeAsString,
  HTTPResponseCode: // KnowSheet HTTPResponseCode,
  HTTPResponse: // KnowSheet HTTPResponse,
  URL: // KnowSheet URL,
  __when1009: 
   { [Function: when]
     promise: [Function: promise],
     resolve: [Function: resolve],
     reject: [Function: reject],
     lift: [Function: lift],
     try: [Function: attempt],
     attempt: [Function: attempt],
     iterate: [Function: iterate],
     unfold: [Function: unfold],
     join: [Function: join],
     all: [Function: all],
     settle: [Function: settle],
     any: [Function],
     some: [Function],
     race: [Function],
     map: [Function: map],
     filter: [Function: filter],
     reduce: [Function],
     reduceRight: [Function],
     isPromiseLike: [Function: isPromiseLike],
     Promise: 
      { [Function: Promise]
        resolve: [Function: resolve],
        reject: [Function: reject],
        never: [Function: never],
        _defer: [Function: defer],
        _handler: [Function: getHandler],
        all: [Function: all],
        race: [Function: race],
        _traverse: [Function: traverse],
        _visitRemaining: [Function: visitRemaining],
        onFatalRejection: [Function],
        onPotentiallyUnhandledRejectionHandled: [Function],
        onPotentiallyUnhandledRejection: [Function],
        exitContext: [Function],
        enterContext: [Function],
        createContext: [Function],
        any: [Function: any],
        some: [Function: some],
        settle: [Function: settle],
        map: [Function: map],
        filter: [Function: filter],
        reduce: [Function: reduce],
        reduceRight: [Function: reduceRight],
        iterate: [Function: iterate],
        unfold: [Function: unfold] },
     defer: [Function: defer],
     TimeoutError: [Function: TimeoutError] } }
1 | (JSONParse(__when1009(HTTP(GET("http://httpbin.org/get?test=1"))).then(function(ret) {
2 |   return ret.body;
3 | }))
4 | )

--------------------------------------------------------------------------------
ReferenceError: GET is not defined
    at [object Object]:1:28
    at Script.(anonymous function) [as runInContext] (vm.js:41:22)
    at Function.evaluate (/home/dima/github/sompylasar/KnowSheet-REPL/lib/bricks-evaluate.js:201:23)
    at KnowSheetREPLServer.options.eval (/home/dima/github/sompylasar/KnowSheet-REPL/lib/repl.js:62:12)
    at Interface.<anonymous> (repl.js:239:12)
    at Interface.emit (events.js:95:17)
    at Interface._onLine (readline.js:202:10)
    at Interface._line (readline.js:531:8)
    at Interface._ttyWrite (readline.js:760:14)
    at ReadStream.onkeypress (readline.js:99:10)
--------------------------------------------------------------------------------

```

sompylasar commented:
@dkorolev Really, cannot reproduce. What `node --version`?

```
$ node --version
v0.12.0
```

sompylasar commented:
Mine:

```
$ NODE_ENV=development node run
KnowSheet> JSONParse(HTTP(GET("http://httpbin.org/get?test=1")).body)
{ DTRACE_NET_SERVER_CONNECTION: undefined,
  DTRACE_NET_STREAM_END: undefined,
  DTRACE_NET_SOCKET_READ: undefined,
  DTRACE_NET_SOCKET_WRITE: undefined,
  DTRACE_HTTP_SERVER_REQUEST: undefined,
  DTRACE_HTTP_SERVER_RESPONSE: undefined,
  DTRACE_HTTP_CLIENT_REQUEST: undefined,
  DTRACE_HTTP_CLIENT_RESPONSE: undefined,
  global: undefined,
  process: undefined,
  GLOBAL: undefined,
  root: undefined,
  Buffer: undefined,
  setTimeout: undefined,
  setInterval: undefined,
  clearTimeout: undefined,
  clearInterval: undefined,
  setImmediate: undefined,
  clearImmediate: undefined,
  console: undefined,
  Array: undefined,
  Boolean: undefined,
  Date: undefined,
  Function: undefined,
  Number: undefined,
  Object: undefined,
  RegExp: undefined,
  String: undefined,
  ArrayBuffer: undefined,
  DataView: undefined,
  Float32Array: undefined,
  Float64Array: undefined,
  Int16Array: undefined,
  Int32Array: undefined,
  Int8Array: undefined,
  Uint16Array: undefined,
  Uint32Array: undefined,
  Uint8Array: undefined,
  Error: undefined,
  EvalError: undefined,
  RangeError: undefined,
  ReferenceError: undefined,
  SyntaxError: undefined,
  TypeError: undefined,
  URIError: undefined,
  require: undefined,
  module: undefined,
  JSON: // KnowSheet JSON,
  JSONParse: // KnowSheet JSONParse,
  SaveResponseToFile: // KnowSheet SaveResponseToFile,
  KeepResponseInMemory: // KnowSheet KeepResponseInMemory,
  HTTP: // KnowSheet HTTP,
  POSTFromFile: // KnowSheet POSTFromFile,
  POST: // KnowSheet POST,
  GET: // KnowSheet GET,
  HTTPResponseCodeAsString: // KnowSheet HTTPResponseCodeAsString,
  HTTPResponseCode: // KnowSheet HTTPResponseCode,
  HTTPResponse: // KnowSheet HTTPResponse,
  URL: // KnowSheet URL,
  __when1128: 
   { [Function: when]
     promise: [Function: promise],
     resolve: [Function: resolve],
     reject: [Function: reject],
     lift: [Function: lift],
     try: [Function: attempt],
     attempt: [Function: attempt],
     iterate: [Function: iterate],
     unfold: [Function: unfold],
     join: [Function: join],
     all: [Function: all],
     settle: [Function: settle],
     any: [Function],
     some: [Function],
     race: [Function],
     map: [Function: map],
     filter: [Function: filter],
     reduce: [Function],
     reduceRight: [Function],
     isPromiseLike: [Function: isPromiseLike],
     Promise: 
      { [Function: Promise]
        resolve: [Function: resolve],
        reject: [Function: reject],
        never: [Function: never],
        _defer: [Function: defer],
        _handler: [Function: getHandler],
        all: [Function: all],
        race: [Function: race],
        _traverse: [Function: traverse],
        _visitRemaining: [Function: visitRemaining],
        onFatalRejection: [Function],
        onPotentiallyUnhandledRejectionHandled: [Function],
        onPotentiallyUnhandledRejection: [Function],
        exitContext: [Function],
        enterContext: [Function],
        createContext: [Function],
        any: [Function: any],
        some: [Function: some],
        settle: [Function: settle],
        map: [Function: map],
        filter: [Function: filter],
        reduce: [Function: reduce],
        reduceRight: [Function: reduceRight],
        iterate: [Function: iterate],
        unfold: [Function: unfold] },
     defer: [Function: defer],
     TimeoutError: [Function: TimeoutError] } }
1 | JSONParse(__when1128(HTTP(GET("http://httpbin.org/get?test=1"))).then(function(ret) {
2 |   return ret.body;
3 | }))
4 | 

--------------------------------------------------------------------------------
{ args: { test: '1' },
  headers: { Host: 'httpbin.org' },
  origin: '95.27.144.243',
  url: 'http://httpbin.org/get?test=1' }
--------------------------------------------------------------------------------
(508ms)
KnowSheet> 

```

dkorolev commented:
My node is `v0.10.33`.

Let me look in more details tomorrow, falling asleep for the past half an hour already. Will look at the commits you pushed into this pull request first thing in the morning.

Thanks,
Dima

sompylasar commented:
Please update your node first. Thanks.

sompylasar commented on lib/bricks-evaluate.js:
In JavaScript, there is no per-scope auto-destruction.

dkorolev commented on lib/bricks-evaluate.js:
Yes, but it makes `UnRegister()` the only void-returning method.

dkorolev commented:
Works fine on node `v0.12.0`!

dkorolev commented:
I looked through the code, it appears your extra output is only there when `NODE_ENV=development` is set, and thus the code is production-ready, and thus it is safe to merge. :+1: 

Thanks!
Dima

sompylasar commented on lib/bricks-evaluate.js:
This applies more to methods which should output nothing after evaluation. The "nothing" does not always mean there was an error.

sompylasar commented:
Great! I will add the "engines" requirement to package.json

dkorolev commented on lib/bricks-evaluate.js:
If it gets tricky, we can always wrap `C++`-s `void` into a special type that indicates that `no news is good news`.

sompylasar commented on lib/bricks-evaluate.js:
Exactly. That is how I did the special pretty-printing for Documentation.

dkorolev commented on lib/bricks-evaluate.js:
:+1:
